File: dotnet/xmscommon/framing/ContentHeaderBody.cs
using System;
using jpmorgan.mina.common;
using log4net;

namespace OpenAMQ.Framing
{
    public class ContentHeaderBody : IBody
    {
        public static readonly byte TYPE = 2;

        public ushort ClassId;

        public ushort Weight;
        
        public ulong BodySize;

        /** must never be null */
        public IContentHeaderProperties Properties;

        public ContentHeaderBody()
        {
        }

        public ContentHeaderBody(IContentHeaderProperties props, ushort classId)
        {
            Properties = props;
            ClassId = classId;
        }

        public ContentHeaderBody(ushort classId, ushort weight, IContentHeaderProperties props, uint bodySize)
            : this(props, classId)
        {            
            Weight = weight;
            BodySize = bodySize;
        }

        #region IBody Members

        public byte BodyType
        {
            get
            {
                return TYPE;
            }
        }

        public uint Size
        {
            get
            {
                return (2 + 2 + 8 + 2 + Properties.PropertyListSize);
            }
        }

        public void WritePayload(ByteBuffer buffer)
        {
            buffer.Put(ClassId);
            buffer.Put(Weight);
            buffer.Put(BodySize);
            buffer.Put(Properties.PropertyFlags);
            Properties.WritePropertyListPayload(buffer);
        }

        public void PopulateFromBuffer(ByteBuffer buffer, uint size)
        {     
            ClassId = buffer.GetUnsignedShort();
            Weight = buffer.GetUnsignedShort();
            BodySize = buffer.GetUnsignedLong();
            ushort propertyFlags = buffer.GetUnsignedShort();
            ContentHeaderPropertiesFactory factory = ContentHeaderPropertiesFactory.GetInstance();
            Properties = factory.CreateContentHeaderProperties(ClassId, propertyFlags, buffer);    
        }

        #endregion

        public static AMQFrame CreateAMQFrame(ushort channelId, ushort classId, ushort weight, BasicContentHeaderProperties properties,
                                              uint bodySize)
        {
            AMQFrame frame = new AMQFrame();
            frame.Channel = channelId;
            frame.BodyFrame = new ContentHeaderBody(classId, weight, properties, bodySize);
            return frame;
        }

        public static AMQFrame CreateAMQFrame(ushort channelId, ContentHeaderBody body)
        {
            AMQFrame frame = new AMQFrame();
            frame.Channel = channelId;
            frame.BodyFrame = body;
            return frame;
        }
    }
}

File: java/src/org/openamq/framing/ContentHeaderBody.java
package org.openamq.framing;

import org.apache.mina.common.ByteBuffer;

public class ContentHeaderBody extends AMQBody
{
    public static final byte TYPE = 2;

    public int classId;

    public int weight;

    /** unsigned long but java can't handle that anyway when allocating byte array */
    public long bodySize;

    /** must never be null */
    public ContentHeaderProperties properties;

    public ContentHeaderBody()
    {
    }

    public ContentHeaderBody(ContentHeaderProperties props, int classId)
    {
        properties = props;
        this.classId = classId;
    }

    public ContentHeaderBody(int classId, int weight, ContentHeaderProperties props, long bodySize)
    {
        this(props, classId);
        this.weight = weight;
        this.bodySize = bodySize;
    }

    protected byte getType()
    {
        return TYPE;
    }

    protected void populateFromBuffer(ByteBuffer buffer, long size) throws AMQFrameDecodingException
    {
        classId = buffer.getUnsignedShort();
        weight = buffer.getUnsignedShort();
        bodySize = buffer.getLong();
        int propertyFlags = buffer.getUnsignedShort();
        ContentHeaderPropertiesFactory factory = ContentHeaderPropertiesFactory.getInstance();
        properties = factory.createContentHeaderProperties(classId, propertyFlags, buffer, (int)size - 14);
    }

    /**
     * Helper method that is used currently by the persistence layer (by BDB at the moment).
     * @param buffer
     * @param size
     * @return
     * @throws AMQFrameDecodingException
     */
    public static ContentHeaderBody createFromBuffer(ByteBuffer buffer, long size) throws AMQFrameDecodingException
    {
        ContentHeaderBody body = new ContentHeaderBody();
        body.populateFromBuffer(buffer, size);
        return body;
    }

    public int getSize()
    {
        return 2 + 2 + 8 + 2 + properties.getPropertyListSize();
    }

    public void writePayload(ByteBuffer buffer)
    {
        EncodingUtils.writeUnsignedShort(buffer, classId);
        EncodingUtils.writeUnsignedShort(buffer, weight);
        buffer.putLong(bodySize);
        EncodingUtils.writeUnsignedShort(buffer, properties.getPropertyFlags());
        properties.writePropertyListPayload(buffer);
    }

    public static AMQFrame createAMQFrame(int channelId, int classId, int weight, BasicContentHeaderProperties properties,
                                          long bodySize)
    {
        final AMQFrame frame = new AMQFrame();
        frame.channel = channelId;
        frame.bodyFrame = new ContentHeaderBody(classId, weight, properties, bodySize);
        return frame;
    }

    public static AMQFrame createAMQFrame(int channelId, ContentHeaderBody body)
    {
        final AMQFrame frame = new AMQFrame();
        frame.channel = channelId;
        frame.bodyFrame = body;
        return frame;
    }
}
